File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.c
// Problem: https://leetcode.cn/problems/longest-substring-without-repeating-characters/
// Accepted at: 2024年5月4日 13:05

int lengthOfLongestSubstring(char* s) {
    int length=strlen(s);
    int hash[128]={0};
    int maxlen=0;
    for(int i=0,j=0;i<length;i++){
        hash[s[i]]++;
        while(hash[s[i]] > 1) hash[s[j++]]--;
        maxlen = maxlen < (i-j+1)?(i-j+1):maxlen;
    }
    return maxlen;
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
// Problem: https://leetcode.cn/problems/longest-substring-without-repeating-characters/
// Accepted at: 2026年6月1日 17:26

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let maxLen = 0;
    let left = 0, right = 0;
    const map = new Map();
    while (right < s.length) {
        if (map.has(s[right])) {
            left = Math.max(left, map.get(s[right]) + 1);
        }
        map.set(s[right], right);
        maxLen = Math.max(maxLen, right - left + 1);
        right++;
    }
    return maxLen;
};
